activesupport/lib/active_support/current_attributes.rb

```diff
@@ -0,0 +1,201 @@
+  #     def create
+  #       Current.account.messages.create(message_params)
+  #     end
+  #   end
+  #
+  #   class Message < ApplicationRecord
+  #     belongs_to :creator, default: -> { Current.user }
+  #     after_create { |message| Event.create(record: message) }
+  #   end
+  #
+  #   class Event < ApplicationRecord
+  #     before_create do
+  #       self.request_id = Current.request_id
+  #       self.user_agent = Current.user_agent
+  #       self.ip_address = Current.ip_address
+  #     end
+  #   end
+  #
+  # A word of caution: It's easy to overdo a global singleton like Current and tangle your model as a result.
+  # Current should only be used for a few, top-level globals, like account, user, and request details.
+  # The attributes stuck in Current should be used by more or less all actions on all requests. If you start
+  # sticking controller-specific attributes in there, you're going to create a mess.
+  class CurrentAttributes
+    include ActiveSupport::Callbacks
+    define_callbacks :reset
+
+    class << self
+      # Returns singleton instance for this class in this thread. If none exists, one is created.
+      def instance
+        Thread.current[:"current_attributes_for_#{name}"] ||= new.tap do |instance|
+          current_instances << instance
+        end
+      end
+
+      # Declares one or more attributes that will be given both class and instance accessor methods.
+      def attribute(*names)
+        generated_attribute_methods.module_eval do
+          names.each do |name|
+            define_method(name) do
+              attributes[name.to_sym]
+            end
+
+            define_method("#{name}=") do |attribute|
+              attributes[name.to_sym] = attribute
+            end
+          end
+        end
+
+        names.each do |name|
+          define_singleton_method(name) do
+            instance.public_send(name)
+          end
+
+          define_singleton_method("#{name}=") do |attribute|
+            instance.public_send("#{name}=", attribute)
+          end
+        end
+      end
+
+      # Delegate methods on both the class and instance level. Example:
+      #
+      #   class Current < ActiveSupport::CurrentAttributes
+      #     attribute :user
+      #     bidelegate :identity, to: :user
+      #   end
+      #
+      #   Current.instance.identity == Current.identity
+      def bidelegate(*methods, to:)
+        methods.each do |method|
+          define_method(method)           { |*args| public_send(to).public_send(method, *args) }
+          define_singleton_method(method) { |*args| public_send(to).public_send(method, *args) }
+        end
+      end
+
+      # Calls this block after #reset is called on the instance. Used for resetting external collaborators, like Time.zone.
+      def resets(&block)
+        set_callback :reset, :after, &block
+      end
+
+      delegate :set, :reset, to: :instance
+
+      def reset_all # :nodoc:
+        current_instances.each(&:reset)
```

Comment:
We don't need to clear, since the `Thread.current` stash for the one instance isn't nulled out too. The instances are meant to be kept around and reset.